content/11-serverless/03-serverless-challenges.mdx

```diff
@@ -0,0 +1,105 @@
+
+## Application design constraints
+
+Another challenge that is fundamental to the serverless model is the application design it imposes.  Serverless platforms are only useful for applications that can work within their constraints.  Some of these are inherent in cloud computing in general, while other requirements are dictated by the serverless model specifically.
+
+### Designing a cloud-friendly architecture
+
+The first requirement that applications must meet to use serverless platforms is to be designed in a cloud-friendly way.  In the context of this discussion, at the very least this means that the application must be at least partly deployable to a cloud service that the other components are able to communicate with.  And while it's possible to implement monolithic functions in your design, the serverless model best accommodates microservice architectures.
+
+The upshot of this is that your application must be designed in part as a series of functions executed by your serverless provider.  You must be comfortable with the processing taking place on infrastructure that you do not control.  Furthermore, you must be able to decompose your application's functionality into discrete functions that can be executed remotely.
+
+### Dealing with stateless execution
+
+Serverless functions are, by design, stateless.  That means that, while some information may possibly be cached if the function is executed with the same resources, you can't rely on any state being shared between invocations of your functions.
+
+You must design your functions to have all of the information they need to execute internally.  Any external state must be fetched at the beginning of invocation and exported before finishing.  Since functions may be executed in parallel, this also limits what type of state may reasonable acted upon.  In general, the less state that your functions have to manage, the faster and cheaper they will be to execute and the less complexity you will have to manage.
+
+There are other side effects of the function's ephemeral nature as well.  If your functions need to reach out to a database system, there is a good chance you may quickly exhaust your database's connection pool.  Since each invocation of your functions can be executed in a different context, your database's connection pool can quickly drain as it responds to different invocations or tries to return resources to its pool.  Solutions like [Prism's Data Proxy](https://www.prisma.io/docs/concepts/components/prisma-data-platform#prisma-data-proxy) help mitigate these issues by managing the connection resources for the serverless instances in front of whatever connection pooling is in place.
+
+## Provider lock-in concerns
+
+One challenge that is difficult to get away from with serverless is provider lock-in.  When you architect your application to rely on external functions running on a specific provider's platform, it can be difficult to migrate to a different platform at a later time.
```

Comment:
We mention "vendor lock-in" in the glossary. I can update the glossary to match or vice versa. Up to you what's better in this context.